adapters/richaudience/richaudience.go

```diff
@@ -26,36 +26,85 @@ func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters
 }
 
 func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
-	raiHeaders := http.Header{}
+	var requestDataRequest []*adapters.RequestData
+
+	errs := make([]error, 0, len(request.Imp))
 
+	raiHeaders := http.Header{}
 	setHeaders(&raiHeaders)
 
 	isUrlSecure := getIsUrlSecure(request)
 
-	resImps, err := setImp(request, isUrlSecure)
-	if err != nil {
-		return nil, []error{err}
-	}
+	for _, imp := range request.Imp {
+		var secure = int8(0)
+		raiExt, err := parseImpExt(&imp)
+		if err != nil {
+			errs = append(errs, &errortypes.BadInput{
+				Message: err.Error(),
+			})
+			return nil, errs
+		}
 
-	request.Imp = resImps
+		if raiExt != nil {
+			if raiExt.Pid != "" {
+				imp.TagID = raiExt.Pid
+			}
 
-	if err = validateDevice(request); err != nil {
-		return nil, []error{err}
-	}
+			if raiExt.Test {
+				request.Device.IP = "11.222.33.44"
```

Comment:
It seems that we should move the `validateDevice(request)` call before this line in order to avoid a potential `nil` pointer de-reference. Maybe even before the `parseImpExt(&imp)` call given that `validateDevice` is less resorce-consuming than the unmarshal operations that `parseImpExt(&imp)` performs and equally makes the function return with an error.
```
28   func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
29       var requestDataRequest []*adapters.RequestData
30   *--  7 lines: errs := make([]error, 0, len(request.Imp))---------------------------
37  
38       for _, imp := range request.Imp {
39           var secure = int8(0)
   +         if err = validateDevice(request); err != nil {
   +             errs = append(errs, &errortypes.BadInput{
   +                 Message: err.Error(),
   +             })
   +             return nil, errs
   +         }
40           raiExt, err := parseImpExt(&imp)
41           if err != nil {
42               errs = append(errs, &errortypes.BadInput{
43                   Message: err.Error(),
44               })
45               return nil, errs
46           }
47  
48           if raiExt != nil {
49               if raiExt.Pid != "" {
50                   imp.TagID = raiExt.Pid
51               }
52  
53               if raiExt.Test {
54                   request.Device.IP = "11.222.33.44"
55                   request.Test = int8(1)
56               }
57  
58               if raiExt.BidFloorCur != "" {
59                   imp.BidFloorCur = raiExt.BidFloorCur
60               } else if imp.BidFloorCur == "" {
61                   imp.BidFloorCur = "USD"
62               }
63           }
64   *-- 17 lines: if isUrlSecure {-----------------------------------------------------
81           request.Imp = []openrtb2.Imp{imp}
82  
83 -         if err = validateDevice(request); err != nil {
84 -             errs = append(errs, &errortypes.BadInput{
85 -                 Message: err.Error(),
86 -             })
87 -             return nil, errs
88 -         }
89  
90           req, err := json.Marshal(request)
91           if err != nil {
92               errs = append(errs, &errortypes.BadInput{
adapters/richaudience/richaudience.go
```
Thougts?

Reply:
Also, can we add a JSON test where `raiExt.Test` is set to `true` and `request.Device` is `nil`? Just to assert we got this scenario segfault free

Reply:
I have changed the position of function, and add the testCase for the scenario

Reply:
@richaudience I'm very sorry for this confusion, I got the location of the `validateDevice(request)` wrong in my previous comment. I meant to make the `validateDevice(request)` call from outside the for loop because there's no need to execute this function more than once since it doesn't perform any checks on the `request.Imp` elements. What do you think?
```
 28   func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
 29       var requestDataRequest []*adapters.RequestData
 30  
 31       errs := make([]error, 0, len(request.Imp))
 32  
 33       raiHeaders := http.Header{}
 34       setHeaders(&raiHeaders)
 35  
 36       isUrlSecure := getIsUrlSecure(request)
 37  
    +     if err := validateDevice(request); err != nil {
    +         errs = append(errs, &errortypes.BadInput{
    +             Message: err.Error(),
    +         })
    +         return nil, errs
    +     }
    +        
 38       for _, imp := range request.Imp {
 39           var secure = int8(0)
 40  
 41 -         if err := validateDevice(request); err != nil {
 42 -             errs = append(errs, &errortypes.BadInput{
 43 -                 Message: err.Error(),
 44 -             })
 45 -             return nil, errs
 46 -         }
 47  
 48           raiExt, err := parseImpExt(&imp)
 49           if err != nil {
 50               errs = append(errs, &errortypes.BadInput{
 51                   Message: err.Error(),
 52               })
 53               return nil, errs
 54           }
 55  
 56   *-- 43 lines: if raiExt != nil {-------------------------------------------------------------------------------------------
 99           requestDataRequest = append(requestDataRequest, &adapters.RequestData{
100               Method:  "POST",
101               Uri:     a.endpoint,
102               Body:    req,
103               Headers: raiHeaders,
104           })
105  
106       }
107  
108       return requestDataRequest, nil
109   }
adapters/richaudience/richaudience.go
```